avocado/core/requirements/resolver.py

```diff
@@ -0,0 +1,106 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2020
+# Authors: Willian Rampazzo <willianr@redhat.com>
+
+"""
+Requirements Resolver module.
+"""
+
+from ..nrunner import RUNNERS_REGISTRY_PYTHON_CLASS, Runnable, Task
+
+SUPPORTED_REQUIREMENTS = {}
+
+
+def create_requirements_tasks(tasks):
+    """Entry point for the Requirements Resolver.
+
+    This function isolates the Requirements Resolver object into this module.
+
+    :param tasks: List of tasks
+    :type tasks: list
+    """
+    resolver = RequirementsResolver(tasks)
+    return resolver.create_requirements_tasks()
```

Comment:
I am a bit confused about this approach. Why a separate entry point into a class with the same method name? Is there a reason to keep the class within the module? It doesn't seem like a private class from the notation. I see two simpler ways to do this, either import the class in the other modules like
```python
from resolver import RequirementsResolver
```
and then use it normally or importing a single `create_requirements_tasks` module-level function which could be this one here. From what I read about the class `RequirementsResolver` below it seems to be nothing but a class with one method returning an extended list of tasks which might as well simply be this method here. Is there some nonobvious reason for the entire `RequirementsResolver` class and the current increased complexity?

Reply:
> I am a bit confused about this approach. Why a separate entry point into a class with the same method name? Is there a reason to keep the class within the module? It doesn't seem like a private class from the notation. I see two simpler ways to do this, either import the class in the other modules like
> 
> ```python
> from resolver import RequirementsResolver
> ```
> 
> and then use it normally or importing a single `create_requirements_tasks` module-level function which could be this one here. 

I find this approach counter-intuitive and, depending on how it is implemented, may become more complex to maintain. Anyway, as you and @beraldoleal opposed to it, I'll change that.

> From what I read about the class `RequirementsResolver` below it seems to be nothing but a class with one method returning an extended list of tasks which might as well simply be this method here. Is there some nonobvious reason for the entire `RequirementsResolver` class and the current increased complexity?

Indeed, this is a simple dispatcher class, just like described in the BP002. This is meant to reduce the complexity in the future addition of new requirements types. So, my implementation here is thinking about extensibility in the future. Without a dispatcher and with multiple requirements types, I don't see how it could be simpler.

Reply:
Can you provide more details on how it will reduce the future complexity of new requirement types? Let's say we have `PackageRequirement` and `ImageRequirement` for instance, how would helping extra dispatcher class help here?

Reply:
In BP002 (https://avocado-framework.readthedocs.io/en/latest/blueprints/BP002.html), we specified 6 types of requirements that we are planning to support: file, image, package, git repository, source code, operating system parameters. We already discussed the requirements resolver could evolve to support environment requirements, like a specific VM, a specific host type with a special configuration in a cloud farm.

I see two options for removing the dispatcher class:
- Move the code to the already existing entry point function `create_requirements_tasks`. It would work for me without problems, but you and @beraldoleal are not happy with that function.
- Move the code inside the `suite.py` file and remove the call to the `create_requirements_tasks`. In my opinion, this will mix the requirements resolver code with the suite code and make things harder to maintain. Imagine I need to change something in the way the requirements classes are called. I would expect to change a code inside the requirements resolver file, but actually, I would need to dig the call on the `suite.py` file.

So, besides these two options, what solution would you suggest to handle 6+ different types of requirements?

Reply:
I am still not sure how the current dispatcher class helps this extensibility but what I meant before was that I am fine with either keeping this in either directly calling the `create_requirements_tasks` methods from the resolver class (rather than having a duplicate module level function) or directly calling a single module function like this one without having an entire dispatcher class just to pass around the list. So I am fine with your point 1 above, what I was pointing out earlier is the middle solution where you have both the module and the class `create_requirements_tasks` while one of them seems to be enough.

Reply:
Okay, got it. Thanks for clarifying!